web/src/main/java/org/springframework/security/web/access/DelegatingMissingAuthorityAccessDeniedHandler.java

```diff
@@ -0,0 +1,217 @@
+		AuthorizationDeniedException denied = findAuthorizationDeniedException(ex);
+		if (denied == null) {
+			return List.of();
+		}
+		if (!(denied.getAuthorizationResult() instanceof AuthorityAuthorizationDecision authorization)) {
+			return List.of();
+		}
+		return authorization.getAuthorities();
+	}
+
+	private @Nullable AuthorizationDeniedException findAuthorizationDeniedException(AccessDeniedException ex) {
+		if (ex instanceof AuthorizationDeniedException denied) {
+			return denied;
+		}
+		Throwable[] chain = this.throwableAnalyzer.determineCauseChain(ex);
+		return (AuthorizationDeniedException) this.throwableAnalyzer
+			.getFirstThrowableOfType(AuthorizationDeniedException.class, chain);
+	}
+
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	/**
+	 * A builder for configuring the set of authority/entry-point pairs
+	 *
+	 * @author Josh Cummings
+	 * @since 7.0
+	 */
+	public static final class Builder {
+
+		private final Map<String, DelegatingAuthenticationEntryPoint.Builder> entryPointByRequestMatcherByAuthority = new LinkedHashMap<>();
+
+		private Builder() {
+
+		}
+
+		DelegatingAuthenticationEntryPoint.Builder entryPointBuilder(String authority) {
+			return this.entryPointByRequestMatcherByAuthority.computeIfAbsent(authority,
+					(k) -> DelegatingAuthenticationEntryPoint.builder());
+		}
+
+		void entryPoint(String authority, AuthenticationEntryPoint entryPoint) {
+			DelegatingAuthenticationEntryPoint.Builder builder = DelegatingAuthenticationEntryPoint.builder()
+				.addEntryPointFor(entryPoint, AnyRequestMatcher.INSTANCE);
+			this.entryPointByRequestMatcherByAuthority.put(authority, builder);
+		}
+
+		/**
+		 * Bind these authorities to the given {@link AuthenticationEntryPoint}
+		 * @param entryPoint the {@link AuthenticationEntryPoint} for the given
+		 * authorities
+		 * @param authorities the authorities
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder addEntryPointFor(AuthenticationEntryPoint entryPoint, String... authorities) {
+			for (String authority : authorities) {
+				Builder.this.entryPoint(authority, entryPoint);
+			}
+			return this;
+		}
+
+		/**
+		 * Bind these authorities to the given {@link AuthenticationEntryPoint}
+		 * @param entryPoint a consumer to configure the underlying
+		 * {@link DelegatingAuthenticationEntryPoint}
+		 * @param authorities the authorities
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder addEntryPointFor(Consumer<DelegatingAuthenticationEntryPoint.Builder> entryPoint,
+				String... authorities) {
+			for (String authority : authorities) {
+				entryPoint.accept(Builder.this.entryPointBuilder(authority));
+			}
+			return this;
+		}
+
+		public DelegatingMissingAuthorityAccessDeniedHandler build() {
+			Map<String, AuthenticationEntryPoint> entryPointByAuthority = new LinkedHashMap<>();
+			for (String authority : this.entryPointByRequestMatcherByAuthority.keySet()) {
```

Comment:
Consider iterating over `entrySet()` instead so that an additional lookup of the value is not required.